File: game_sfml/v7.cpp
#include <SFML/Graphics.hpp>
#include <iostream>

/// Globals
sf::RenderWindow* window;
int RES_X = 800;
int RES_Y = 600;

bool key_up = false;
bool key_down = false;
bool key_left = false;
bool key_right = false;

class Player;
class Bullet;
class Enemy;
Player* player;
std::vector<Bullet> bullets;
std::vector<Enemy> enemies;
unsigned long long frameCount;


class Bullet {
public:
    int radius = 8;
    bool isActive = true;
    int vx, vy;
    int x, y;

    Bullet(int x, int y, int vx, int vy): x(x),y(y),vx(vx),vy(vy) {}

    void update() {
        if (!isActive) return;

        x += vx;
        y += vy;
        if (x >= RES_X + radius || x < -radius ||
            y >= RES_Y + radius || y < -radius) isActive = false;
    }

    void draw() {
        if (!isActive) return;

        sf::CircleShape shape;
        shape.setFillColor(sf::Color::Yellow);
        shape.setRadius(radius);
        shape.setOrigin(radius,radius);
        shape.setPosition(x,y);
        window->draw(shape);
    }
};

class Player {
public:
    bool isActive = true;
    int width = 40, height = 40;
    int speed = 6;
    int x, y;

    Player(int x, int y): x(x), y(y) {}

    void update() {
        if (!isActive) return;
        
        int vx = 0, vy = 0;
        if (key_up) vy -= speed;
        if (key_down) vy += speed;
        if (key_left) vx -= speed;
        if (key_right) vx += speed;
        x += vx;
        y += vy;
    }

    void draw() {
        if (!isActive) return;
        
        sf::RectangleShape shape;
        shape.setFillColor(sf::Color::Blue);
        shape.setSize(sf::Vector2f(width, height));
        shape.setOrigin(width/2, height/2);
        shape.setPosition(x, y);
        window->draw(shape);
    }
    
    void shoot() {
        if (!isActive) return;
        
        bullets.push_back(Bullet(x,y,20,0));
    }
};


class Enemy {
public:
    float speed = 2.5f;
    const int radius = 60;
    bool isActive = true;
    float x, y;

    Enemy(int x, int y): x(x), y(y) {}

    void update() {
        if (!isActive) return;

        float vx = player->x - x;
        float vy = player->y - y;
        float magnitude = sqrt(vx*vx+vy*vy);
        if (magnitude != 0) {
            vx *= speed / magnitude;
            vy *= speed / magnitude;
        }
        x += vx;
        y += vy;
    }

    void collisionCheck(Bullet& bullet) {
        if (!isActive || !bullet.isActive) return;

        int dx = x - bullet.x;
        int dy = y - bullet.y;
        int r = radius + bullet.radius;
        if (dx*dx+dy*dy <= r*r) {
            isActive = false;
            bullet.isActive = false;
        }
    }

    void collisionCheck(Player* _player) {
        if (!isActive || !_player->isActive) return;

        int dx = x - _player->x;
        int dy = y - _player->y;
        int r = radius;
        if (dx*dx+dy*dy <= r*r) {
            _player->isActive = false;
        }
    }

    void draw() {
        if (!isActive) return;

        sf::CircleShape shape;
        shape.setFillColor(sf::Color::Red);
        shape.setRadius(radius);
        shape.setOrigin(radius,radius);
        shape.setPosition(x,y);
        window->draw(shape);
    }
};

void spawnEnemy() {
    int x = RES_X + 100;
    int y = rand()%RES_Y;
    enemies.push_back(Enemy(x,y));
}


void initialiseGame() {
    player = new Player(RES_X/2, RES_Y/2);
    bullets.clear();
    enemies.clear();
    frameCount = 0;
}

void updateGame() {
    player->update();
    for (int i=0;i<bullets.size();++i) {
        bullets[i].update();
    }
    for (int i=0;i<enemies.size();++i) {
        enemies[i].update();
    }
    
    if (frameCount%30 == 0) {
        spawnEnemy();
    }
    ++frameCount;
}

void drawGameFrame() {
    player->draw();
    for (int i=0;i<bullets.size();++i) {
        bullets[i].draw();
    }
    for (int i=0;i<enemies.size();++i) {
        enemies[i].draw();
        
        for (int j=0;j<bullets.size();++j) {
            enemies[i].collisionCheck(bullets[j]);
        }
        enemies[i].collisionCheck(player);
    }
}

void keyDown(sf::Keyboard::Key keyCode) {
    switch(keyCode) {
        case sf::Keyboard::Up: key_up = true; break;
        case sf::Keyboard::Down: key_down = true; break;
        case sf::Keyboard::Left: key_left = true; break;
        case sf::Keyboard::Right: key_right = true; break;
        case sf::Keyboard::Z: {
            player->shoot();
            break;
        }
        case sf::Keyboard::R: {
            initialiseGame();
            break;
        }
    }
}

void keyUp(sf::Keyboard::Key keyCode) {
    switch(keyCode) {
        case sf::Keyboard::Up: key_up = false; break;
        case sf::Keyboard::Down: key_down = false; break;
        case sf::Keyboard::Left: key_left = false; break;
        case sf::Keyboard::Right: key_right = false; break;
    }
}

void processEvent(sf::Event& event) {
    switch(event.type) {
        case sf::Event::Closed: {
            window->close();
            break;
        }
        case sf::Event::KeyPressed: {
            keyDown(event.key.code);
            break;
        }
        case sf::Event::KeyReleased: {
            keyUp(event.key.code);
            break;
        }
    }
}

int main() {
    window = new sf::RenderWindow(sf::VideoMode(RES_X, RES_Y), "Awesome Game");

    sf::Clock clock;
    float frameTime = 1/60.0f;
    float dTime = 0;

    initialiseGame();

    while (window->isOpen()) {
        dTime += clock.getElapsedTime().asSeconds();
        clock.restart();

        // Event handling
        sf::Event event;
        while(window->pollEvent(event)) {
            processEvent(event);
        }

        // Safeguard (slowdown) to prevent game from lagging to death
        if (dTime > 5*frameTime) dTime = 5*frameTime;

        // Update game
        while (dTime > frameTime) {
            dTime -= frameTime;
            updateGame();
        }

        // Draw frame
        window->clear();
        drawGameFrame();
        window->display();
    }

    delete window;
    delete player;
    return 0;
}

File: game_pygame/v7.py
import sys, pygame
import random
FRAME_UPDATE_EVENT = pygame.USEREVENT+1

screen = None
RES_X = 800
RES_Y = 600

key_up = False
key_down = False
key_left = False
key_right = False

player = None
bullets = []
enemies = []
frameCount = 0

class Bullet(object):
    def __init__(self, x, y, vx, vy):
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.isActive = True
        self.radius = 8

    def update(self):
        if not self.isActive: return
        
        self.x += self.vx
        self.y += self.vy

        if self.x > RES_X + self.radius or self.x < -self.radius or\
           self.y > RES_Y + self.radius or self.y < -self.radius:
           self.isActive = False

    def draw(self):
        if not self.isActive: return
        
        pygame.draw.circle(screen, (255,255,0), (self.x, self.y), self.radius)


    
class Player(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 40
        self.height = 40
        self.speed = 6
        self.isActive = True
        
    def update(self):
        if not self.isActive: return
    
        vx = 0; vy = 0
        if (key_up): vy -= self.speed
        if (key_down): vy += self.speed
        if (key_left): vx -= self.speed
        if (key_right): vx += self.speed
        self.x += vx
        self.y += vy
    
    def draw(self):
        if not self.isActive: return
        
        pygame.draw.rect(screen, (0,0,255),
            (self.x-self.width//2, self.y-self.height//2, self.width, self.height))
            
    def shoot(self):
        if not self.isActive: return
        
        bullets.append(Bullet(self.x,self.y,20,0))
        

class Enemy(object):
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.speed = 2.5
        self.radius = 60
        self.isActive = True
        
    def update(self):
        if not self.isActive: return
        
        vx = player.x - self.x
        vy = player.y - self.y
        magnitude = (vx*vx+vy*vy)**0.5
        if magnitude != 0:
            vx *= self.speed / magnitude
            vy *= self.speed / magnitude
        self.x += vx
        self.y += vy
        
    def bulletCollisionCheck(self, bullet):
        if not self.isActive or not bullet.isActive: return
        
        dx = self.x - bullet.x
        dy = self.y - bullet.y
        r = self.radius + bullet.radius
        if dx*dx+dy*dy <= r*r:
            self.isActive = False
            bullet.isActive = False
            
    def playerCollisionCheck(self, _player):
        if not self.isActive or not _player.isActive: return
        
        dx = self.x - _player.x
        dy = self.y - _player.y
        r = self.radius
        if dx*dx+dy*dy <= r*r:
            _player.isActive = False
        
    def draw(self):
        if not self.isActive: return
        
        pygame.draw.circle(screen, (255,0,0),
            (int(self.x), int(self.y)), self.radius)
        

def spawnEnemy():
    x = RES_X + 100
    y = random.randrange(0,RES_Y)
    enemies.append(Enemy(x,y))


def initialiseGame():
    global player, bullets, enemies, frameCount
    frameCount = 0
    player = Player(RES_X//2, RES_Y//2)
    bullets = []
    enemies = []


def update():
    global frameCount
    player.update()
    for bullet in bullets:
        bullet.update()
    for enemy in enemies:
        enemy.update()
        for bullet in bullets:
            enemy.bulletCollisionCheck(bullet)
        enemy.playerCollisionCheck(player)

    if frameCount % 40 == 0:
        spawnEnemy()
    frameCount += 1


def draw():
    player.draw()
    for bullet in bullets:
        bullet.draw()
    for enemy in enemies:
        enemy.draw()
    

def keyDown(key):
    global key_up, key_down, key_left, key_right
    if key == pygame.K_UP: key_up = True
    elif key == pygame.K_DOWN: key_down = True
    elif key == pygame.K_LEFT: key_left = True
    elif key == pygame.K_RIGHT: key_right = True
    elif key == pygame.K_z: player.shoot()
    elif key == pygame.K_r: initialiseGame()
    

def keyUp(key):
    global key_up, key_down, key_left, key_right
    if key == pygame.K_UP: key_up = False
    elif key == pygame.K_DOWN: key_down = False
    elif key == pygame.K_LEFT: key_left = False
    elif key == pygame.K_RIGHT: key_right = False

def processEvent(event):
    if event.type == FRAME_UPDATE_EVENT:
        update()
    if event.type == pygame.QUIT:
        sys.exit()
    elif event.type == pygame.KEYDOWN:
        keyDown(event.key)
    elif event.type == pygame.KEYUP:
        keyUp(event.key)


def main():
    pygame.init()
    global screen
    screen = pygame.display.set_mode((RES_X, RES_Y))
    
    frameTime = 1000//60 # in ms
    pygame.time.set_timer(FRAME_UPDATE_EVENT, frameTime)
    
    initialiseGame()
    while True:
        for event in pygame.event.get():
            processEvent(event)
        
        screen.fill((0,0,0))
        draw()
        pygame.display.flip()

if __name__ == '__main__':
    main()